components/tikv_util/src/timer.rs

```diff
@@ -196,28 +188,91 @@ impl Default for SteadyTimer {
 }
 
 fn start_global_steady_timer() -> SteadyTimer {
-    let (tx, rx) = mpsc::channel();
     let clock = SteadyClock::default();
-    let clock_ = clock.clone();
+    let handle = start_timer_thread("steady-timer", clock.clone());
+    SteadyTimer { clock, handle }
+}
+
+/// A clock that ratchets forward.
+///
+/// It is used to workaround a panic[^1] in tokio-timer when the clock goes
+/// backward, which is possible in some environments, e.g., aliyun ECS g7.
+///
+/// [^1]: https://github.com/tokio-rs/tokio/pull/515
+struct RatchetClock<T: Now> {
+    clock: T,
+    start: std::time::Instant,
+
+    // The last recorded time in milliseconds.
+    //
+    // Using u64 is sufficient here because tokio-timer operates with millisecond
+    // precision, and `u64::MAX` milliseconds represent an extremely large time span
+    // (several million years).
+    last_now_ms: AtomicU64,
+}
+
+impl<T: Now> Now for RatchetClock<T> {
+    fn now(&self) -> std::time::Instant {
+        let now = self.clock.now();
+        let now_ms = now.saturating_duration_since(self.start).as_millis() as u64;
+        let mut last_now_ms = self.last_now_ms.load(AtomicOrdering::Relaxed);
+        loop {
+            if now_ms > last_now_ms {
+                match self.last_now_ms.compare_exchange_weak(
+                    last_now_ms,
+                    now_ms,
+                    AtomicOrdering::SeqCst,
+                    AtomicOrdering::SeqCst,
+                ) {
+                    Ok(_) => {
+                        return self.start + Duration::from_millis(now_ms);
+                    }
+                    Err(last_now) => {
+                        last_now_ms = last_now;
+                    }
+                }
+            } else {
+                return self.start + Duration::from_millis(last_now_ms);
+            }
+        }
+    }
+}
+
+impl<T: Now> RatchetClock<T> {
+    fn new(clock: T) -> Self {
+        RatchetClock {
+            clock,
+            start: std::time::Instant::now(),
+            last_now_ms: AtomicU64::new(0),
+        }
+    }
+}
+
+/// Start a timer thread with specific thread name and clock source.
+fn start_timer_thread<T: Now>(name: &str, clock: T) -> Handle {
+    let (tx, rx) = mpsc::channel();
+    let props = crate::thread_group::current_properties();
+    let ratchet_clock = RatchetClock::new(clock);
+    let clock = Clock::new_with_now(ratchet_clock);
     Builder::new()
-        .name(thd_name!("steady-timer"))
+        .name(thd_name!(name))
         .spawn_wrapper(move || {
-            let c = Clock::new_with_now(clock_);
-            let mut timer = tokio_timer::Timer::new_with_now(ParkThread::new(), c);
+            crate::thread_group::set_properties(props);
+
+            let mut timer = tokio_timer::Timer::new_with_now(ParkThread::new(), clock);
```

Comment:
I if understand correctly, the timer thread is a single-threaded, it's seems unnecessary to let the clock to be thread safe.

Reply:
The `Now` trait requires `Sync + Send + 'static`, so it has to be thread-safe.

Reply:
I think in this use case, the `Sync` bound is too strong and unnecessary if the clock is only used in the timer thread.

BTW, in the current implementation, you suppose the cloud can be visited by different threads concurrently. If this is true, then the comparison of time value from different threads is meaningless as they don't have a causal order.

Reply:
Although the `RatchetClock` has to be thread-safe, the `Clock` that wraps it can only be accessed by the timer thread. So the `RatchetClock` will not be accessed by multiple threads.

Note that the `Now` trait is defined by the tokio-timer crate, and yes, the thread-safety requirement does seem an unnecessary in the current usage.